sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -361,8 +361,15 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
       case r @ logical.Range(start, end, step, numSlices, output) =>
         execution.Range(start, step, numSlices, r.numElements, output) :: Nil
       case logical.RepartitionByExpression(expressions, child, nPartitions) =>
-        execution.Exchange(HashPartitioning(
-          expressions, nPartitions.getOrElse(numPartitions)), planLater(child)) :: Nil
+        val newPartitioning = HashPartitioning(expressions, nPartitions.getOrElse(numPartitions))
+        val childPlan = self.sqlContext.planner.planLater(child)
+        // This is necessary to get the outputPartitioning from EnsureRequirements batch
+        val executedChildPlan = sqlContext.prepareForExecution.execute(childPlan)
```

Comment:
Oh, I see, because you need `EnsureRequirements` to have run.  Thats actually pretty confusing.  In that case, I think this version (46e741933d9737fb7e85f731da8fb8982e603ab7) is the cleanest implementation, but I'd add a comment about why we would both remove and add them.  Sorry for the back and forth.